mergify_engine/squash_pull.py

```diff
@@ -0,0 +1,132 @@
+# -*- encoding: utf-8 -*-
+#
+#  Copyright © 2021 Mergify SAS
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+import dataclasses
+
+from mergify_engine import context
+from mergify_engine.clients import http
+from mergify_engine.user_tokens import UserTokensUser
+from mergify_engine import gitter
+from mergify_engine import config
+
+
+@dataclasses.dataclass
+class SquashFailure(Exception):
+    reason: str
+
+
+async def _do_squash(ctxt: context.Context, user: UserTokensUser, title: str) -> None:
+
+    head_repo = (
+        ctxt.pull["head"]["repo"]["owner"]["login"]
+        + "/"
+        + ctxt.pull["head"]["repo"]["name"]
+    )
+    base_repo = (
+        ctxt.pull["base"]["repo"]["owner"]["login"]
+        + "/"
+        + ctxt.pull["base"]["repo"]["name"]
+    )
+
+    head_branch = ctxt.pull["head"]["ref"]
+    base_branch = ctxt.pull["base"]["ref"]
+
+    git = gitter.Gitter(ctxt.log)
+
+    try:
+        await git.init()
+        # NOTE(sileht): Bump the repository format. This ensures required
+        # extensions (promisor, partialclonefilter) are present in git cli and
+        # raise an error if not. Avoiding git cli to fallback to full clone
+        # behavior for us.
+        await git("config", "core.repositoryformatversion", "1")
+
+        if ctxt.subscription.active:
+            await git.configure(user["name"] or user["login"], user["email"])
+        else:
+            await git.configure()
+        await git.add_cred(user["oauth_access_token"], "", head_repo)
+        await git.add_cred(user["oauth_access_token"], "", base_repo)
+
+        await git("remote", "add", "origin", f"{config.GITHUB_URL}/{head_repo}")
+        # https://github.com + /mergify-test2/head_branch
```

Comment:
Can you add the following config to the remote, to avoid fetch the whole repository:
```
        await git("config", "remote.origin.promisor", "true")
        await git("config", "remote.origin.partialclonefilter", "blob:none")
```